apps/emqx/src/emqx_persistent_session_ds_state.erl

```diff
@@ -0,0 +1,582 @@
+%% distributed under the License is distributed on an "AS IS" BASIS,
+%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+%% See the License for the specific language governing permissions and
+%% limitations under the License.
+%%--------------------------------------------------------------------
+
+%% @doc CRUD interface for the persistent session
+%%
+%% This module encapsulates the data related to the state of the
+%% inflight messages for the persistent session based on DS.
+%%
+%% It is responsible for saving, caching, and restoring session state.
+%% It is completely devoid of business logic. Not even the default
+%% values should be set in this module.
+-module(emqx_persistent_session_ds_state).
+
+-export([create_tables/0]).
+
+-export([open/1, create_new/1, delete/1, commit/1, format/1, print_session/1, list_sessions/0]).
+-export([get_created_at/1, set_created_at/2]).
+-export([get_last_alive_at/1, set_last_alive_at/2]).
+-export([get_conninfo/1, set_conninfo/2]).
+-export([new_id/1]).
+-export([get_stream/2, put_stream/3, del_stream/2, fold_streams/3]).
+-export([get_seqno/2, put_seqno/3]).
+-export([get_rank/2, put_rank/3, del_rank/2, fold_ranks/3]).
+-export([get_subscriptions/1, put_subscription/4, del_subscription/3]).
+
+-export([make_session_iterator/0, session_iterator_next/2]).
+
+-export_type([
+    t/0, metadata/0, subscriptions/0, seqno_type/0, stream_key/0, rank_key/0, session_iterator/0
+]).
+
+-include("emqx_mqtt.hrl").
+-include("emqx_persistent_session_ds.hrl").
+-include_lib("snabbkaffe/include/trace.hrl").
+-include_lib("stdlib/include/qlc.hrl").
+
+%%================================================================================
+%% Type declarations
+%%================================================================================
+
+-type subscriptions() :: emqx_topic_gbt:t(_SubId, emqx_persistent_session_ds:subscription()).
+
+-opaque session_iterator() :: emqx_persistent_session_ds:id() | '$end_of_table'.
+
+%% Generic key-value wrapper that is used for exporting arbitrary
+%% terms to mnesia:
+-record(kv, {k, v}).
+
+%% Persistent map.
+%%
+%% Pmap accumulates the updates in a term stored in the heap of a
+%% process, so they can be committed all at once in a single
+%% transaction.
+%%
+%% It should be possible to make frequent changes to the pmap without
+%% stressing Mria.
+%%
+%% It's implemented as three maps: `clean', `dirty' and `tombstones'.
+%% Updates are made to the `dirty' area. `pmap_commit' function saves
+%% the updated entries to Mnesia and moves them to the `clean' area.
+-record(pmap, {table, cache, dirty}).
+
+-type pmap(K, V) ::
+    #pmap{
+        table :: atom(),
+        cache :: #{K => V},
+        dirty :: #{K => dirty | del}
+    }.
+
+%% Session metadata:
+-define(created_at, created_at).
+-define(last_alive_at, last_alive_at).
+-define(conninfo, conninfo).
+%% Unique integer used to create unique identities
+-define(last_id, last_id).
```

Comment:
Translation of a map from/to a compact representation will be done here: https://github.com/emqx/emqx/pull/12251/files#diff-4f926b90b373b839679213fbdb3c30c784f5575bc3d4b721afce21466dded77eR394 (I haven't implemented any encoder callbacks yet).

I found that integer-keyed maps are rather unpleasant for debugging, I'd like to avoid them when possible.